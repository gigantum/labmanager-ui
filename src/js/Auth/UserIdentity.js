//vendor
import {
  graphql,
} from 'react-relay'
//mutations
import RemoveUserIdentityMutation from 'Mutations/RemoveUserIdentityMutation'
//environment
import {fetchQuery} from 'JS/createRelayEnvironment';

const userIdentityQuery = graphql`
  query UserIdentityQuery{
    userIdentity{
      id
      username
      email
      givenName
      familyName
      isSessionValid
    }
  }
`;

const UserIdentity = {
  getUserIdentity: () => {

    return new Promise((resolve, reject) =>{

      let fetchData = function(){

        fetchQuery(userIdentityQuery(), {}).then((response, error) => {

          if(response){
            resolve(response)
          }else{

            reject(response)
          }
        }).catch((error) =>{

          reject(error)
        })
      }

      fetchData()
    })
  },
  removeUserIdentity: () => {
    RemoveUserIdentityMutation(()=>{
      console.log('removed user')
    })
  }

}

export default UserIdentity
