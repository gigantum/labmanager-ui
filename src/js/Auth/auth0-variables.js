export const AUTH_CONFIG = {
  domain: 'calum.auth0.com',
  clientId: '9wFN5CZA4Z9xa1r8KZIZbevpjHG7bAab',
  callbackUrl: window.location.origin + '/callback'
}
