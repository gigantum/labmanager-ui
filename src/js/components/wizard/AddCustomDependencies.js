//vendor
import React from 'react'
import SweetAlert from 'sweetalert-react';
import { QueryRenderer, graphql } from 'react-relay'
//components
import Loader from 'Components/shared/Loader'
//utilites
import environment from 'JS/createRelayEnvironment'
//mutations
import AddEnvironmentComponentMutation from 'Mutations/AddEnvironmentComponentMutation'
import BuildImageMutation from 'Mutations/BuildImageMutation'

const AddCustomDependenciesQuery = graphql`query AddCustomDependenciesQuery($first: Int!, $cursor: String){
  availableCustomDependencies(first: $first, after: $cursor){
    edges{
      node{
        id
        component{
          id
          repository
          namespace
          name
          version
          componentClass
        }
        author{
          id
          name
          email
          username
          organization
        }
        info{
          id
          name
          humanName
          description
          versionMajor
          versionMinor
          tags
          icon
        }
        osBaseClass
        docker
      }
      cursor
    }
    pageInfo{
      endCursor
      hasNextPage
      hasPreviousPage
      startCursor
    }
  }
}`

export default class AddCustomDependencies extends React.Component {
  constructor(props){
  	super(props);
  	this.state = {
      'modal_visible': false,
      'selectedCustomDependencies': [],
      'selectedCustomDependenciesIds': [],
      'show': false,
      'message': '',
      'isLoading': false
    };
    this.continueSave = this.continueSave.bind(this)
    this._buildLabbook = this._buildLabbook.bind(this)
    this._selectCustomDependency = this._selectCustomDependency.bind(this)
    this._addCustomDependencies = this._addCustomDependencies.bind(this)
  }

  /**
    @param {Object} edge
    takes a base image edge
    sets componest state for selectedCustomDependencyId and selectedCustomDependency
  */
  _selectCustomDependency = (edge) => {
    let selectedCustomDependencies = this.state.selectedCustomDependencies;
    let selectedCustomDependenciesIds = this.state.selectedCustomDependenciesIds;

    if(selectedCustomDependenciesIds.indexOf(edge.node.id) > -1){
        selectedCustomDependenciesIds.splice(selectedCustomDependenciesIds.indexOf(edge.node.id), 1)
        selectedCustomDependencies.splice(selectedCustomDependenciesIds.indexOf(edge.node.id), 1)
    }else{
      selectedCustomDependenciesIds.push(edge.node.id)
      selectedCustomDependencies.push(edge)
    }
    this.setState({'selectedCustomDependencies': selectedCustomDependencies})
    this.setState({'selectedCustomDependenciesIds': selectedCustomDependenciesIds})
    this.props.toggleDisabledContinue(false);
  }

  /**
    @param {}
    triggers buildMuation
    sends user to next window
  */
  _buildLabbook = () => {

    BuildImageMutation(
      this.props.labbookName,
      'default',
      (response, error) => {
        console.log(response, error)
        let showAlert = ((error !== undefined) && (error !== null))
        if(showAlert){
          let message = showAlert ? error[0].message : '';
          this.setState({
            'isLoading': false,
            'show': showAlert,
            'message': message
          })
        }
        if(this.props.buildCallback){
          this.props.buildCallback()
        }
        this.props.setComponent(this.props.nextWindow);

      })
  }


  /**
    @param {}
    gets current selectedCustomDependency and passes variables to AddEnvironmentComponentMutation
    callback triggers build mutation
  */
  _addCustomDependencies = () => {
    let all = [];

    this.setState({
      'isLoading': true
    })

    this.props.toggleDisabledContinue(true);

    this.state.selectedCustomDependencies.forEach((edge) => {

      const {component} = edge.node;

      let promise = new Promise((resolve, reject) => {
        AddEnvironmentComponentMutation(
          this.props.labbookName,
          'default',
          component.repository,
          component.namespace,
          component.name,
          component.version,
          "clientMutationId",
          this.props.environmentId,
          this.props.connection,
          component.componentClass,
          (error) => {
            console.log(error)
            let showAlert = (error)

            if(showAlert){
              let message = showAlert ? error[0].message : '';
              this.setState({
                'show': showAlert,
                'message': message,
                'reject': reject

              })

            }else{

              resolve()
            }
          }
        )
      })

      all.push(promise)
    });

    Promise.all(all).then(values => {
        this._buildLabbook();
    })
  }


  /**
    @param {boolean} isSkip
    gets current selectedCustomDependency and passes variables to AddEnvironmentComponentMutation
    callback triggers and modal state is changed to  next window
  */
  continueSave(isSkip){
    if(!isSkip){
      this._addCustomDependencies()
    }
    else{
      this._buildLabbook()
    }
  }

  /**
    @param {}
    @return {Object} environmentView object
  */
  _environmentView(){
    return this.props.environmentView
  }

  render(){

    return(
      <div className="AddCustomDependencies">

        <p> Select Custom Dependencies </p>

        <QueryRenderer
          variables={{
            first: 20
          }}
          query={AddCustomDependenciesQuery}
          environment={environment}
          render={({error, props}) =>{

              if(error){

                return(<div>{error.message}</div>)
              }else{

                if(props){
                  return(
                    this._renderQueryResults(props)

                  )
                }else{
                  return(<Loader />)
                }
              }
          }}
        />

      </div>
      )
  }


  _renderQueryResults(props){
    return(
      <div className="AddCustomDependencies__inner-container flex flex--column justify--space-between">
        <div className="AddCustomDependencies__selected-image-container">

            {
              (this.state.selectedCustomDependencies.length > -1) && (
                this.state.selectedCustomDependencies.map((customDependency, index) => {
                  return (
                    <div key={customDependency.node.id + index} className="AddCustomDependencies__selected-image flex">
                      <img alt="" src={customDependency.node.info.icon} height="50" width="50" />
                      <p>{customDependency.node.info.humanName}</p>
                    </div>
                  )
                })
              )
            }
        </div>

        <div className="AddCustomDependencies__images flex flex--row flex--wrap justify--space-around">
        {
          props.availableCustomDependencies.edges.map((edge) => {

              let disabled = (this.state.selectedCustomDependenciesIds.indexOf(edge.node.id) > -1)

              return(
                <div
                  disabled={disabled}
                  className={disabled ? 'AddCustomDependencies__image--selected': 'AddCustomDependencies__image'}
                  onClick={()=> this._selectCustomDependency(edge)}
                  key={edge.node.id}>

                  <img alt="" src={edge.node.info.icon} height="50" width="50" />
                  <p>{edge.node.info.humanName}</p>

                </div>
              )
          })
        }

      </div>
      {
        this._environmentView() && (
          <div className="SelectBaseImage__progress-buttons flex flex--row justify--space-between">
            <button className="SelectBaseImage__save-button" onClick={() => this.continueSave()}>Save</button>
          </div>
        )
      }

      <div className={this.state.isLoading ? '' : 'hidden'}>
        <Loader />
      </div>

      <SweetAlert
        className="sa-error-container"
        show={this.state.show}
        type="error"
        title="Error"
        text={this.state.message}
        onConfirm={() => {
          this.state.reject(); this.setState({ show: false, message: ''})
        }}
        />
    </div>
    )
  }
}
