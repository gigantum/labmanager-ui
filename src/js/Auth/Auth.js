import history from 'JS/history';
import auth0 from 'auth0-js';
import { AUTH_CONFIG } from './auth0-variables';
import RemoveUserIdentityMutation from 'Mutations/RemoveUserIdentityMutation'
//store
import store from 'JS/redux/store'


export default class Auth {
  auth0 = new auth0.WebAuth({
    domain: AUTH_CONFIG.domain,
    clientID: AUTH_CONFIG.clientId,
    redirectUri: AUTH_CONFIG.callbackUrl,
    audience:  AUTH_CONFIG.audience,
    responseType: 'token id_token',
    scope: 'openid profile email user_metadata'
  });

  constructor() {
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
    this.handleAuthentication = this.handleAuthentication.bind(this);
    this.isAuthenticated = this.isAuthenticated.bind(this);

  }

  login() {
    this.auth0.authorize();
  }

  handleAuthentication() {

    this.auth0.parseHash((err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {

        this.setSession(authResult);

      } else if (err) {
        console.error(err);
      //  alert(`Error: ${err.error}. Check the console for further details.`);
      }
    });
  }

  setSession(authResult) {

    // Set the time that the access token will expire at
    let expiresAt = JSON.stringify((authResult.expiresIn * 1000) + new Date().getTime());
    localStorage.setItem('access_token', authResult.accessToken);
    localStorage.setItem('id_token', authResult.idToken);
    localStorage.setItem('expires_at', expiresAt);
    localStorage.setItem('family_name', authResult.idTokenPayload.family_name);
    localStorage.setItem('given_name', authResult.idTokenPayload.given_name);
    localStorage.setItem('email', authResult.idTokenPayload.email);
    localStorage.setItem('username', authResult.idTokenPayload.nickname);
    //redirect to labbooks when user logs in

    let storrageRoute = sessionStorage.getItem('CALLBACK_ROUTE')
    let route = storrageRoute !== '' ? storrageRoute : `/labbooks`

    history.replace(route)
  }

  logout() {

    localStorage.removeItem('access_token');
    localStorage.removeItem('id_token');
    localStorage.removeItem('expires_at');
    localStorage.removeItem('family_name')
    localStorage.removeItem('given_name');
    localStorage.removeItem('email');
    localStorage.removeItem('username');
    RemoveUserIdentityMutation(()=>{
      //redirect to root when user logs out
      history.replace('/');
    })
  }

  isAuthenticated() {
    // Check whether the current time is past the
    // access token's expiry time
    let expiresAt = JSON.parse(localStorage.getItem('expires_at'));

    return new Date().getTime() < expiresAt;
  }
}
