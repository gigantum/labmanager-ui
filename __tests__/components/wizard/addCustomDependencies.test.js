import React from 'react';
import AddCustomDependencies from 'Components/wizard/AddCustomDependencies';
import {mount} from 'enzyme'
import renderer from 'react-test-renderer';
import Auth from 'JS/Auth/Auth';
import sinon from 'sinon'
import relayTestingUtils from 'relay-testing-utils'
import json from './__relaydata__/AddCustomDependencies.json'


let variables = {first: 5}

export default variables

let toggleDisabledContinue = () => {}
let setComponent = () => {}

test('Test AddCustomDependencies rendering', async () => {
  console.log(relayTestingUtils.relayWrap)
  const component = await renderer.create(
    relayTestingUtils.relayWrap(
      <AddCustomDependencies
        data={json.data}
        availableCustomDependencies={json.data.availableCustomDependencies}
        toggleDisabledContinue={toggleDisabledContinue}
        setComponent={setComponent}
        nextWindow={'addDatasets'}
        labbookName={'demo-lab-book'}/>, {}, json.data
      )
  );


  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});

//
// describe('Test AddCustomDependencies rendering', () => {
//
//   const component = mount(
//     relayTestingUtils.relayWrap(<AddCustomDependencies
//
//       toggleDisabledContinue={toggleDisabledContinue}
//       setComponent={setComponent}
//       nextWindow={'addDatasets'}
//       labbookName={'demo-lab-book'}/>)
//   );
//   console.log(component)
//
// });
