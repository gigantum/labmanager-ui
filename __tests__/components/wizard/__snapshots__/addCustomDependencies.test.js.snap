// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AddCustomDependencies rendering 1`] = `
<div
  className="AddCustomDependencies"
>
  <p>
     Select Custom Dependencies 
  </p>
</div>
`;
